### PR TITLE
Update to Babel 7.4 and core-js 3

### LINKED ISSUE #15138: Scripts: Missing corejs version in Babel config
**Describe the bug**
I am trying to use `@wordpress/scripts` for unit tests. Still running into #13364, but there's also another issue when running `wp-scripts test-unit-js`:

```
WARNING: We noticed you're using the `useBuiltIns` option without declaring a core-js version. Currently, we assume version 2.x when no version is passed. Since this default version will likely change in future versions of Babel, we recommend explicitly setting the core-js version you are using via the `corejs` option.

You should also be sure that the version you pass to the `corejs` option matches the version specified in your `package.json`'s `dependencies` section. If it doesn't, you need to run one of the following commands:

  npm install --save core-js@2    npm install --save core-js@3
  yarn add core-js@2              yarn add core-js@3
```

Indeed, `@wordpress/babel-preset-default` uses `useBuiltIns` when running tests, but does not specify the `corejs` version:

https://github.com/WordPress/gutenberg/blob/354c1ff6b5337837eb4183d4eb8302ffeccd1be8/packages/babel-preset-default/index.js#L20-L21

**To reproduce**
1. Install `wp-scripts` in your project.
2. Try to run some tests
3. Notice error

**Expected behavior**

The Babel config should add the `corejs` option in that scenario.

### PR DESCRIPTION
**Pull request by swissspidy**
## Description

Specifies the version of `core-js` used in the package (Gutenberg uses 2.x at the moment) to prevent warnings.

## How has this been tested?

1. Install `wp-scripts` in your project.
2. Try to run some tests
3. Notice that there's no error anymore.

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 

Fixes #15138.

### CODE REVIEW

**Comment by gziolo** [comment]
It breaks some of the unit tests related to our Babel preset and one of the Babel plugins:

https://travis-ci.com/WordPress/gutenberg/jobs/195206255#L5966-L6221

**Comment by nerrad** [comment]
The full warning @swissspidy is referring to is:

```
WARNING: We noticed you're using the `useBuiltIns` option without declaring a core-js version. Currently, we assume version 2.x when no version is passed. Since this default version will likely change in future versions of Babel, we recommend explicitly setting the core-js version you are using via the `corejs` option.

You should also be sure that the version you pass to the `corejs` option matches the version specified in your `package.json`'s `dependencies` section. If it doesn't, you need to run one of the following commands:

  npm install --save core-js@2    npm install --save core-js@3
  yarn add core-js@2              yarn add core-js@3
```

I'm think this warning was introduced with a specific version of @babel/core so if one newer than what @wordpress/packages install is used by the parent package this warning starts showing up.  I started the same warning in Event Espresso when I updated our @babel/core package to 7.4.3.  

I think we should fix the tests to allow this as a top-level option and possibly (if needed) update babel.

**Comment by swissspidy** [comment]
It is indeed a valid top-level option since Babel 7.4.

Related: https://github.com/babel/babel/issues/9751 / https://github.com/babel/babel/pull/7646

**Comment by gziolo** [comment]
Should we update to Babel 7.4.x and `core-js@3` instead? :)

**Comment by swissspidy** [comment]
If it doesn't cause any issues, sure. Let me try it in this PR :-)

**Comment by nerrad** [comment]
A good reference: https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md

**Comment by swissspidy** [comment]
I am currently getting this error when running tests:

```
Cannot find module 'core-js/modules/web.dom.iterable' from 'gutenberg-phase.js'
```

The module has been renamed in `core-js@3`, I just don't know where the old one is still referenced 🤔

**Comment by nerrad** [comment]
I _think_ that's being referenced in the @wordpress/gutenberg-phase package

**Comment by nerrad** [comment]
I think I recall seeing something like this when I attempted to upgrade to corejs3 in the EE repo.  I got as far as uncovering it is related to jest configuration (which would seem to make sense for your error since it happens with a file in the e2e tests).

**Comment by gziolo** [comment]
I think it's an issue with the lock file which installs now many babel related packages as dependencies of dependencies instead of moving them to the top level. I will push fix soon.

**Comment by gziolo** [comment]
With 8c9bc2f applied at least unit tests pass for me locally.

**Comment by gziolo** [comment]
It looks like ESLint isn't happy about new version of Babel:
```
[2] npm run lint-css exited with code 0
[0] TypeError: Cannot read property 'length' of undefined
[0]     at nthChild (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/esquery/esquery.js:256:34)
[0]     at matches (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/esquery/esquery.js:158:25)
[0]     at matches (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/esquery/esquery.js:76:30)
[0]     at matches (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/esquery/esquery.js:103:25)
[0]     at Function.matches (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/esquery/esquery.js:70:29)
[0]     at NodeEventGenerator.applySelector (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/eslint/lib/util/node-event-generator.js:250:21)
[0]     at NodeEventGenerator.applySelectors (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/eslint/lib/util/node-event-generator.js:280:22)
[0]     at NodeEventGenerator.enterNode (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/eslint/lib/util/node-event-generator.js:294:14)
[0]     at CodePathAnalyzer.enterNode (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:632:23)
[0]     at nodeQueue.forEach.traversalInfo (/Users/gziolo/PhpstormProjects/gutenberg/node_modules/eslint/lib/linter.js:750:28)
```

**Comment by gziolo** [comment]
Another try in ca77a9d :)

ESLint still throws an error but I can confirm that e2e tests pass.

**Comment by swissspidy** [comment]
That error looks familiar: https://github.com/estools/esquery/issues/67

**Comment by gziolo** [comment]
> That error looks familiar: estools/esquery#67

@aduth do you know whether the error thrown by ESLint is coming from the custom rule based on of the `no-restricted-syntax` rules?

File: .eslintrc.js
Comment (gziolo) [inline .eslintrc.js]: @aduth - this is the rule that causes ESLint to randomly throw errors when processing JSX code with the latest version of both Babel and ESLint. Any ideas about how to fix it?

**Comment by gziolo** [comment]
47feb53 fixes linting by removing the rule which throws errors. The only remaining issue is fixing e2e tests. I still suspect that this is cache issue on Travis. I can't replicate this error locally.

File: .eslintrc.js
Comment (aduth) [reply .eslintrc.js]: > @aduth - this is the rule that causes ESLint to randomly throw errors when processing JSX code with the latest version of both Babel and ESLint. Any ideas about how to fix it?

I think we avoided the error previously based on the fact that filtering to `Literal` would be applied before calling the `nth-child`, as it would occur when trying to access properties on specific types of nodes. I'm guessing something changed here with the order in how the filtering is applied.

I imagine if we were to move more of this implementation into JavaScript (i.e. not rely on this `esquery` AST selector syntax), it could work again. Not immediately sure how much effort might be involved, or if it's more worthwhile to consider moving it to its own proper custom rule in `@wordpress/eslint-plugin`.

Comment (aduth) [reply .eslintrc.js]: Tried a few ideas briefly, but no luck so far.

As far as I can tell, there's no way to include JavaScript in the rule to implement logic around considering the selected value in `no-restricted-syntax` (reasonable considering ESLint configurations are often JSON). The only options are `selector` and `message`.

I tried a variation which opted for separate rules to test ` > :not(ArrayExpression)` and ` > * + :not(ArrayExpression)` for first and second arguments respectively, but puzzlingly the same error occurs (despite not using `:nth-child`).

```js
{
	selector: (
		`CallExpression[callee.name=/^(${ [
			'property',
			'matchesProperty',
			'path',
		].join( '|' ) })$/] > :not(ArrayExpression)`
	),
	message: 'Always pass an array as the path argument',
},
{
	selector: (
		`CallExpression[callee.name=/^(${ [
			'invokeMap',
			'get',
			'has',
			'hasIn',
			'invoke',
			'result',
			'set',
			'setWith',
			'unset',
			'update',
			'updateWith',
		].join( '|' ) })$/] > * + :not(ArrayExpression)`
	),
	message: 'Always pass an array as the path argument',
},
```

**Comment by gziolo** [comment]
`core-js/modules/web.dom.iterable` - I have no idea where this issue comes from because there is no such file in `node_modules` folder in my local project ...

**Comment by swissspidy** [comment]
`core-js/modules/web.dom.iterable` is the file name from core-js v2. It was changed/renamed in core-js v3. Maybe a cache issue?

**Review by gziolo — APPROVED** [review]
I applied a few commits in this PR so it might make sense to have another reviewer to go through this PR as well :)

I think it looks solid. Tests pass locally and on Travis after issues with the cache and test configs fot resolved. 

As discussed in the comments. One ESLint rule was removed as it seems like there is no way to make it work with the latest versions of Babel and ESLint. I will open a follow-up issue with all the details after this change lands.

**Review by nerrad — APPROVED** [review]
👍 However, should all the packages that received `@babel/runtime` bumps have changelog entries for that bump?

**Comment by gziolo** [comment]
> However, should all the packages that received `@babel/runtime` bumps have changelog entries for that bump?

That's a great question. We can add it, it won't harm :) Technically, it shouldn't be an issue because as I understand `@babel/runtime` should work with both `coreja@2` and `corejs@3`.

**Comment by gziolo** [comment]
> 👍 However, should all the packages that received `@babel/runtime` bumps have changelog entries for that bump?

It looks like `@babel/runtime` update has nothing to do with `core-js` and it's version. It's mostly to bring `regeneratorRuntime`. See related discussion here:
https://github.com/WordPress/gutenberg/pull/9171/files#r214595243

File: packages/babel-preset-default/index.js
Comment (gziolo) [inline packages/babel-preset-default/index.js]: It looks like we enabled `corejs` v3 only for tests which is expected. Related discussion here: https://github.com/WordPress/gutenberg/pull/9171/files#r214595243

File: .eslintrc.js
Comment (aduth) [reply .eslintrc.js]: I'm fine to have this not considered a blocker for merge. The rule is good to have, but not critical, and can be reintroduced in the future.

**Pull request merged**
